src/common/vfs.h

```diff
@@ -30,72 +33,138 @@ class VfsPrivate;
+     */
     virtual void start(const VfsSetupParams &params) = 0;
+
+    /// Stop interaction with VFS provider. Like when the client application quits.
     virtual void stop() = 0;
+
+    /// Deregister the folder with the sync provider, like when a folder is removed.
     virtual void unregisterFolder() = 0;
 
+
+    /** Return true when download of a file's data is currently ongoing.
+     *
+     * See also the beginHydrating() and doneHydrating() signals.
+     */
     virtual bool isHydrating() const = 0;
 
-    // Update placeholder metadata during discovery
+    /** Update placeholder metadata during discovery.
+     *
+     * If the remote metadata changes, the local placeholder's metadata should possibly
+     * change as well.
+     *
+     * Returning false and setting error indicates an error.
+     */
     virtual bool updateMetadata(const QString &filePath, time_t modtime, quint64 size, const QByteArray &fileId, QString *error) = 0;
 
-    // Create and convert placeholders in PropagateDownload
+    /// Create a new dehydrated placeholder. Called from PropagateDownload.
     virtual void createPlaceholder(const QString &syncFolder, const SyncFileItemPtr &item) = 0;
+
+    /** Convert a new file to a hydrated placeholder.
+     *
+     * Some VFS integrations expect that every file, including those that have all
+     * the remote data, are "placeholders". This function is called by PropagateDownload
+     * to convert newly downloaded, fully hydrated files into placeholders.
+     */
     virtual void convertToPlaceholder(const QString &filename, const SyncFileItemPtr &item) = 0;
 
-    // Determine whether something is a placeholder
+    /// Determine whether the file at the given absolute path is a dehydrated placeholder.
     virtual bool isDehydratedPlaceholder(const QString &filePath) = 0;
 
-    // Determine whether something is a placeholder in discovery
-    // stat has at least 'path' filled
-    // the stat_data argument has platform specific data
-    // returning true means that the file_stat->type was set and should be fixed
+    /** Similar to isDehydratedPlaceholder() but used from sync discovery.
+     *
+     * This function shall set stat->type if appropriate.
+     * It may rely on stat->path and stat_data (platform specific data).
+     *
+     * Returning true means that type was fully determined.
+     */
     virtual bool statTypeVirtualFile(csync_file_stat_t *stat, void *stat_data) = 0;
 
 signals:
+    /// Emitted when a user-initiated hydration starts
     void beginHydrating();
+    /// Emitted when the hydration ends
     void doneHydrating();
 };
 
+/// Check whether the plugin for the mode is available.
 OCSYNC_EXPORT bool isVfsPluginAvailable(Vfs::Mode mode);
+
+/// Return the best available VFS mode.
 OCSYNC_EXPORT Vfs::Mode bestAvailableVfsMode();
-OCSYNC_EXPORT Vfs *createVfsFromPlugin(Vfs::Mode mode, QObject *parent);
+
+/// Create a VFS instance for the mode, returns nullptr on failure.
```

Comment:
Maybe instead of returning nullptr on faillure, we should return the VFS instance for Vfs::Off.
Yes, i know currently this does not exist, but maybe it should.
This way we do not need to check if vfs is null all the time.

Reply:
I had a similar thought and mentioned it in one of the commit messages. Having a VfsOff instance and being guaranteed that vfs!=nullptr would make this less error-prone. Will do.